cloudsql/mysql/database-sql/cloudsql.go

```diff
@@ -0,0 +1,299 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Sample database-sql demonstrates connection to a Cloud SQL instance from App Engine
+// standard. The application is a Golang version of the "Tabs vs Spaces" web
+// app presented at Cloud Next '19 as seen in this video:
+// https://www.youtube.com/watch?v=qVgzP3PsXFw&t=1833s
+package main
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"html/template"
+	"log"
+	"math"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/go-sql-driver/mysql"
+)
+
+// db is the global database connection pool.
+var db *sql.DB
+
+// parsedTemplate is the global parsed HTML template.
+var parsedTemplate *template.Template
+
+// vote struct contains a single row from the votes table in the database.
+// Each vote includes a candidate ("TABS" or "SPACES") and a timestamp.
+type vote struct {
+	Candidate string
+	VoteTime  mysql.NullTime
+}
+
+// voteDiff is used to provide a string representation of the current voting
+// margin, such as "1 vote" (singular) or "2 votes" (plural).
+type voteDiff int
+
+func (v voteDiff) String() string {
+	if v == 1 {
+		return "1 vote"
+	}
+	return strconv.Itoa(int(v)) + " votes"
+}
+
+// templateData struct is used to pass data to the HTML template.
+type templateData struct {
+	TabsCount   uint
+	SpacesCount uint
+	VoteMargin  string
+	RecentVotes []vote
+}
+
+func main() {
+	var err error
+
+	parsedTemplate, err = template.ParseFiles("templates/index.html")
+	if err != nil {
+		log.Fatalf("unable to parse template file: %s", err)
+	}
+
+	// If the optional DB_TCP_HOST environment variable is set, it contains
+	// the IP address and port number of a TCP connection pool to be created,
+	// such as "127.0.0.1:3306". If DB_TCP_HOST is not set, a Unix socket
+	// connection pool will be created instead.
+	if os.Getenv("DB_TCP_HOST") != "" {
+		db, err = initTcpConnectionPool()
+		if err != nil {
+			log.Fatalf("initTcpConnectionPool: unable to connect: %s", err)
+		}
+	} else {
+		db, err = initSocketConnectionPool()
+		if err != nil {
+			log.Fatalf("initSocketConnectionPool: unable to connect: %s", err)
+		}
+	}
+
+	// Create the votes table if it does not already exist.
+	if _, err = db.Exec(`CREATE TABLE IF NOT EXISTS votes
+	( vote_id SERIAL NOT NULL, time_cast timestamp NOT NULL,
+	candidate CHAR(6) NOT NULL, PRIMARY KEY (vote_id) );`); err != nil {
+		log.Fatalf("unable to create table: %s", err)
+	}
+
+	http.HandleFunc("/", indexHandler)
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+	}
+
+	log.Printf("Listening on port %s", port)
+	if err := http.ListenAndServe(":"+port, nil); err != nil {
+		log.Fatal(err)
+	}
+
+}
+
+// indexHandler handles requests to the / route.
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		err := showTotals(w, r)
+		if err != nil {
+			log.Fatalf("indexHandler: error returned from showTotals: %s", err)
```

Comment:
1. log.Fatal is a crash, should not be used unless the problem transcends a single request's concerns.
2. More concise error message and pattern:

```suggestion
			log.Printf("showTotals: %v", err)
```